x-pack/apm-server/aggregation/servicemetrics/aggregator.go

```diff
@@ -0,0 +1,342 @@
+		stopped:  make(chan struct{}),
+		config:   config,
+		active:   newMetricsBuffer(config.MaxGroups),
+		inactive: newMetricsBuffer(config.MaxGroups),
+	}, nil
+}
+
+// Run runs the Aggregator, periodically publishing and clearing aggregated
+// metrics. Run returns when either a fatal error occurs, or the Aggregator's
+// Stop method is invoked.
+func (a *Aggregator) Run() error {
+	ticker := time.NewTicker(a.config.Interval)
+	defer ticker.Stop()
+	defer func() {
+		a.stopMu.Lock()
+		defer a.stopMu.Unlock()
+		select {
+		case <-a.stopped:
+		default:
+			close(a.stopped)
+		}
+	}()
+	var stop bool
+	for !stop {
+		select {
+		case <-a.stopping:
+			stop = true
+		case <-ticker.C:
+		}
+		if err := a.publish(context.Background()); err != nil {
+			a.config.Logger.With(logp.Error(err)).Warnf(
+				"publishing service metrics failed: %s", err,
+			)
+		}
+	}
+	return nil
+}
+
+// Stop stops the Aggregator if it is running, waiting for it to flush any
+// aggregated metrics and return, or for the context to be cancelled.
+//
+// After Stop has been called the aggregator cannot be reused, as the Run
+// method will always return immediately.
+func (a *Aggregator) Stop(ctx context.Context) error {
+	a.stopMu.Lock()
+	select {
+	case <-a.stopped:
+	case <-a.stopping:
+		// Already stopping/stopped.
+	default:
+		close(a.stopping)
+	}
+	a.stopMu.Unlock()
+
+	select {
+	case <-a.stopped:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	return nil
+}
+
+func (a *Aggregator) publish(ctx context.Context) error {
+	// We hold a.mu only long enough to swap the serviceMetrics. This will
+	// be blocked by serviceMetrics updates, which is OK, as we prefer not
+	// to block serviceMetrics updaters. After the lock is released nothing
+	// will be accessing a.inactive.
+	a.mu.Lock()
+
+	// EXPLAIN: We swap active <-> inactive, so that we're only working on the
+	// inactive property while publish is running. `a.active` is the buffer that
+	// receives/stores/updates the metricsets, once swapped, we're working on the
+	// `a.inactive` which we're going to process and publish.
+	a.active, a.inactive = a.inactive, a.active
+	a.mu.Unlock()
+
+	size := len(a.inactive.m)
+	if size == 0 {
+		a.config.Logger.Debugf("no service metrics to publish")
+		return nil
+	}
+
+	batch := make(model.Batch, 0, size)
+	for key, metrics := range a.inactive.m {
+		metricset := makeMetricset(key, metrics)
+		batch = append(batch, metricset)
+		delete(a.inactive.m, key)
+	}
+	a.config.Logger.Debugf("publishing %d metricsets", len(batch))
+	return a.config.BatchProcessor.ProcessBatch(ctx, &batch)
+}
+
+// ProcessBatch aggregates all service latency metrics
+func (a *Aggregator) ProcessBatch(ctx context.Context, b *model.Batch) error {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	for _, event := range *b {
+		tx := event.Transaction
+		if event.Processor == model.TransactionProcessor && tx != nil {
+			if msEvent := a.processTransaction(&event); msEvent.Metricset != nil {
```

Comment:
I wonder if this can be extended to also observe `errors` and `metrics` and emits this as `errorCount` and `metricsCount` metrics that way querying for services can be even further simplified:

https://github.com/elastic/kibana/blob/2f7603de181a47cde19eccb6fd4b665954bd105b/x-pack/plugins/apm/server/routes/services/get_services/get_services_items.ts

Today merges two results client side to create a holistic picture.